cmd/kubeadm/app/cmd/upgrade/common_test.go

```diff
@@ -23,6 +23,136 @@ import (
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
 )
 
+func TestGetK8sVersionFromUserInput(t *testing.T) {
+	var tcases = []struct {
+		name               string
+		isVersionMandatory bool
+		flags              *applyPlanFlags
+		args               []string
+		expectedErr        bool
+		expectedVersion    string
+	}{
+		{
+			name:               "No config and version as an argument",
+			isVersionMandatory: true,
+			flags:              &applyPlanFlags{},
+			args:               []string{"v1.13.1"},
+			expectedVersion:    "v1.13.1",
+		},
+		{
+			name:               "Neither config nor version specified",
+			isVersionMandatory: true,
+			flags:              &applyPlanFlags{},
+			expectedErr:        true,
+		},
+		{
+			name:               "No config and empty version as an argument",
+			isVersionMandatory: true,
+			flags:              &applyPlanFlags{},
+			args:               []string{""},
+			expectedErr:        true,
+		},
+		{
+			name:               "Valid config, but no version specified",
+			isVersionMandatory: true,
+			flags: &applyPlanFlags{
+				cfgPath: "testdata/cluster_config_with_version.yaml",
```

Comment:
I am not a big a fan of using testdata files. In fact I was planning on nuking those at some point. Trouble is, that these occasionally get recycled by other tests, the actual test data is far apart from the test code and the test data itself does not mention what it's used to test.
I'd rather keep the content of the config in a const in this file (or even test case), write it out in a temp file and supply the temp file path to `cfgPath`.
But that's just me, I am fine if you choose to keep the testdata files.

Reply:
Rewrote to create temp files with configs.